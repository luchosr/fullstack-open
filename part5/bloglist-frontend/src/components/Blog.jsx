import { useState } from 'react';

import blogService from '../services/blogs';

const Blog = ({ blog, updateView }) => {
  const [toggleDetails, setToggleDetails] = useState(false);

  return (
    <>
      <div
        style={{
          border: '1px solid black',
          width: '50%',
          borderRadius: '5px',
          padding: 5,
          margin: 10,
        }}
      >
        {blog.title} {blog.author}
        <button
          style={{ marginLeft: 5 }}
          type="button"
          onClick={() => setToggleDetails(!toggleDetails)}
        >
          {toggleDetails ? 'hide' : 'view'}
        </button>
      </div>
      {toggleDetails ? (
        <div
          style={{
            border: '1px solid black',
            width: '50%',
            borderRadius: '5px',
            padding: 5,
            margin: 10,
          }}
        >
          <a
            style={{ display: 'block', overflow: 'hidden' }}
            href={blog.url}
            target="_blank"
          >
            {blog.url}
          </a>
          <p>
            Likes: {blog.likes}{' '}
            <button
              style={{ marginLeft: 5 }}
              type="button"
              onClick={() => {
                blogService
                  .update(blog.id, {
                    user: blog.user,
                    likes: blog.likes + 1,
                    author: blog.author,
                    title: blog.title,
                    url: blog.url,
                  })
                  .then(() => updateView());
              }}
            >
              like it!
            </button>
          </p>
          <p>Author: {blog.author}</p>

          <button
            type="button"
            onClick={() => {
              confirm(`remove blog ${blog.title} by ${blog.author}??`)
                ? blogService.deleteId(blog).then(() => updateView())
                : '';
            }}
          >
            Remove
          </button>
        </div>
      ) : (
        ''
      )}
    </>
  );
};

export default Blog;
