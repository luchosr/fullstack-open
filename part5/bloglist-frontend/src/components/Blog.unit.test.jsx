import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Blog from './Blog';

test('renders content', () => {
  const blog = {
    title: 'Component testing is done with react-testing-library',
    author: 'luciano',
    url: 'www.testing.com',
    likes: 4,
  };

  const mockHandler = vi.fn();
  const { container } = render(
    <Blog blog={blog} updateLikes={mockHandler} removeBlog={mockHandler} />
  );
  const div = container.querySelector('.blog');

  expect(div).toHaveTextContent(
    'Component testing is done with react-testing-library'
  );
});

test('blog component should show only a title and an blog author ', () => {
  const blog = {
    title: 'Component testing is done with react-testing-library',
    author: 'luciano',
    url: 'www.testing.com',
    likes: 4,
  };

  const mockHandler = vi.fn();
  const { container } = render(
    <Blog blog={blog} updateLikes={mockHandler} removeBlog={mockHandler} />
  );
  const h3 = container.querySelector('.blog-title');
  const h4 = container.querySelector('.blog-author');
  const details = container.querySelector('.togglableContent');

  expect(h3).toHaveTextContent(
    'Component testing is done with react-testing-library'
  );
  expect(h4).toHaveTextContent('luciano');
  expect(details).toHaveStyle('display: none');
});

//siempre cuando hay un click es asÃ­ncrono (async / await)
test('blog should show blog url and likes when show details button is clicked', async () => {
  const blog = {
    title: 'Component testing is done with react-testing-library',
    author: 'luciano',
    url: 'www.testing.com',
    likes: 4,
  };

  const mockHandler = vi.fn();
  const { container } = render(
    <Blog blog={blog} updateLikes={mockHandler} removeBlog={mockHandler} />
  );

  const user = userEvent.setup();
  const button = screen.getByText('Show details');

  await user.click(button);

  const div = container.querySelector('.togglableContent');
  const detailsUrl = container.querySelector('.details-url');
  const detailsLikes = container.querySelector('.blog-likes');

  expect(div).not.toHaveStyle('display: none');
  expect(detailsUrl).toHaveTextContent('www.testing.com');
  expect(detailsLikes).toHaveTextContent('Likes: 4');
});

test('likes button should two times and fire the amount of likes of the blog', async () => {
  const blog = {
    title: 'Component testing is done with react-testing-library',
    author: 'luciano',
    url: 'www.testing.com',
    likes: 4,
  };
  const mockHandler = vi.fn();
  const { container } = render(
    <Blog blog={blog} updateLikes={mockHandler} removeBlog={mockHandler} />
  );

  const user = userEvent.setup();
  const showDetailsButton = screen.getByText('Show details');

  await user.click(showDetailsButton);

  const detailsLikes = container.querySelector('.blog-likes-button');

  await user.click(detailsLikes);
  await user.click(detailsLikes);

  expect(mockHandler.mock.calls).toHaveLength(2);
});

test('<Blog /> should pass a snapshot testing', () => {
  const blog = {
    title: 'Component testing is done with react-testing-library',
    author: 'luciano',
    url: 'www.testing.com',
    likes: 4,
  };

  const mockHandler = vi.fn();

  const newBlog = render(
    <Blog blog={blog} updateLikes={mockHandler} removeBlog={mockHandler} />
  );

  expect(newBlog).toMatchSnapshot();
});
