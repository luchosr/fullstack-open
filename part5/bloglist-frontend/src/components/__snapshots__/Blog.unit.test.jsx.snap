// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Blog /> should pass a snapshot testing 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="blog"
        style="border: 1px solid black; width: 50%; border-radius: 5px; padding: 5px; margin: 10px;"
      >
        <h3
          class="blog-title"
        >
           
          Component testing is done with react-testing-library
        </h3>
        <h4
          class="blog-author"
        >
           
          luciano
        </h4>
        <div>
          <div>
            <button>
              Show details
            </button>
          </div>
          <div
            class="togglableContent"
            style="display: none;"
          >
            <div
              class="blog-details"
            >
              <a
                class="details-url"
                href="www.testing.com"
                rel="noopener noreferrer"
                style="display: block; overflow: hidden;"
                target="_blank"
              >
                www.testing.com
              </a>
              <p
                class="blog-likes"
              >
                Likes: 
                4
                 
              </p>
              <button
                class="blog-likes-button"
                style="margin-left: 5px;"
                type="button"
              >
                like it!
              </button>
              <p>
                Author: 
                luciano
              </p>
              <button
                type="button"
              >
                Remove
              </button>
            </div>
            <button>
              cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="blog"
      style="border: 1px solid black; width: 50%; border-radius: 5px; padding: 5px; margin: 10px;"
    >
      <h3
        class="blog-title"
      >
         
        Component testing is done with react-testing-library
      </h3>
      <h4
        class="blog-author"
      >
         
        luciano
      </h4>
      <div>
        <div>
          <button>
            Show details
          </button>
        </div>
        <div
          class="togglableContent"
          style="display: none;"
        >
          <div
            class="blog-details"
          >
            <a
              class="details-url"
              href="www.testing.com"
              rel="noopener noreferrer"
              style="display: block; overflow: hidden;"
              target="_blank"
            >
              www.testing.com
            </a>
            <p
              class="blog-likes"
            >
              Likes: 
              4
               
            </p>
            <button
              class="blog-likes-button"
              style="margin-left: 5px;"
              type="button"
            >
              like it!
            </button>
            <p>
              Author: 
              luciano
            </p>
            <button
              type="button"
            >
              Remove
            </button>
          </div>
          <button>
            cancel
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
